kit/ChildSession.cpp

```diff
@@ -868,44 +871,56 @@ bool ChildSession::getStatus(const char* /*buffer*/, int /*length*/)
 
 void ChildSession::dumpRecordedUnoCommands()
 {
-    std::atomic<char*>* recordedCommands = unoCommandsRecorder.getRecordedCommands();
+    // Clean formatiing around the unoCommand dump to see the commands distinct and clear
+    Log::signalLog("\n------------------------------\n");
 
-    for (int i = 0; i < unoCommandsRecorder.NUM_UNO_COMMANDS; i++)
+    for (auto& unoCommandRecorder : UnoCommandsRecorders)
     {
-        // Set the slot to null to prevent other threads from deleting
-        // the command in future( while recording new command)
-        char * unoCommandInfo = recordedCommands[i].exchange(nullptr);
-
-        // If the slot is already empty or signal handlers reentry
-        if(unoCommandInfo != nullptr)
-            Log::signalLog(unoCommandInfo);
+        if (unoCommandRecorder != nullptr)
+        {
+            UnoCommandsRecorder::UnoCommands& unoCommands = unoCommandRecorder->getRecordedCommands();
+            for (auto& unoCommand : unoCommands)
+            {
+                if (unoCommand != nullptr)
+                    Log::signalLog(unoCommand);
+            }
+            Log::signalLog("\n------------------------------\n");
+        }
     }
 }
 
+void ChildSession::registerUnoCommandsRecorder(std::shared_ptr<UnoCommandsRecorder> recorder)
+{
+    UnoCommandsRecorders[CurrentPos] = recorder;
+    CurrentPos = (CurrentPos+1) % NUM_UNO_RECORDERS;
+}
+
+
 UnoCommandsRecorder::UnoCommandsRecorder() :
     _currentpos(0)
 {
     for (int i = 0; i < NUM_UNO_COMMANDS; i++)
     {
-        _unocommands[i].store(nullptr);
+        _unocommands[i] = nullptr;
     }
 }
 
 void UnoCommandsRecorder::addUnoCommandInfo(const std::string& unoCommandInfo)
 {
     // Create a new C string on heap with the contents of unoCommandInfo
-    char* command = new char[unoCommandInfo.length()+1];
-    strncpy(command, unoCommandInfo.c_str(), unoCommandInfo.length()+1);
+    char* newCommand = new char[unoCommandInfo.length()+1];
+    strncpy(newCommand, unoCommandInfo.c_str(), unoCommandInfo.length()+1);
```

Comment:
This is an anti-pattern, I think. strncpy() is hard to use meaningfully and correctly, I tend to avoid it. This is not a meaningful use of strncpy(). You know the length of the source string and you have allocated enough space for it on the previous line, so you could just use strcpy().

There is absolutely no extra "security" in using strncpy() as here. (For instance, if somebody later for some reason reduces the amount of space allocated with the new operator but forgets to change the last parameter to strncpy(), using strncpy() has no benefit, it will gladly overwrite the space allocated for the destination.)

And actually you could just use strdup(), and then deallocate with free().

Reply:
Yes I will update it with strdup.